R/hugo.R

```diff
@@ -142,20 +144,30 @@ install_theme = function(
   theme, hostname = 'github.com', theme_example = FALSE, update_config = TRUE, force = FALSE
 ) {
   r = '^([^/]+/[^/@]+)(@.+)?$'
-  if (!is.character(theme) || length(theme) != 1 || !grepl(r, theme)) {
-    warning("'theme' must be a character string of the form 'user/repo' or 'user/repo@branch'")
+  r_zip = "\\.zip$"
+  theme_is_url = grepl(r_zip, theme)
+  if (!is.character(theme) || length(theme) != 1 || (!grepl(r, theme) & !theme_is_url)) {
+    warning("'theme' must be a character string of the form 'user/repo' or 'user/repo@branch', or a full URL to the .zip file")
     return(invisible())
   }
   branch = sub('^@', '', gsub(r, '\\2', theme))
   if (branch == '') branch = 'master'
   theme = gsub(r, '\\1', theme)
   dir_create('themes')
   in_dir('themes', {
-    url = sprintf('https://%s/%s/archive/%s.zip', hostname, theme, branch)
-    zipfile = sprintf('%s.zip', basename(theme))
+    if (theme_is_url) {
+      url = theme
+      zipfile = gsub(".+/(.+\\.zip)", "\\1", theme)
+    } else {
+      url = sprintf('https://%s/%s/archive/%s.zip', hostname, theme, branch)
+      zipfile = sprintf('%s.zip', basename(theme))
+    }
     download2(url, zipfile, mode = 'wb')
-    files = utils::unzip(zipfile)
+    tmpdir = tempfile("", ".")
+    on.exit(in_dir('themes', unlink(tmpdir, recursive = TRUE)))
+    files = utils::unzip(zipfile, exdir = tmpdir)
     zipdir = dirname(files)
+    zipdir = gsub(tmpdir, ".", zipdir)
```

Comment:
I think this is too early. The dir `expdir` on line 173 won't be found if you strip `tmpdir` from `zipdir` at this point.

Reply:
Good catch! I just moved line 170 down so that it's stripped after finding `expdir`.

Reply:
Actually, on second thought, it's only needed for `newdir`, so I'm going to refactor a bit.